<div>{{jsSidebar("JavaScript Guide")}} {{PreviousNext("Web/JavaScript/Guide/Grammar_and_types", "Web/JavaScript/Guide/Loops_and_iteration")}}</div>

<p class="summary">JavaScript は Web ページに多様なインタラクティビティをもたらすコンパクトな{{原語併記("文","statement")}} の集合体、特に制御フロー文をサポートしています。本章ではこれらの文の概要を説明します。</p>

<p><a href="/ja/docs/Web/JavaScript/Reference/Statements">JavaScript リファレンス</a>にはこの章で紹介する文についての完全な詳細が載っています。また、JavaScript のコードではセミコロン (<code>;</code>) 文字で文を区切ります。</p>

<p>あらゆる JavaScript の式は、文でもあります。式に関する詳細については<a href="/ja/docs/JavaScript/Guide/Expressions_and_Operators">式と演算子</a>を参照ください。</p>

<h2 id="Block_statement" name="Block_statement">ブロック文</h2>

<p>たいていの場合、ブロック文は文のグループ化に使います。ブロックは、1 組の波括弧で区切られます :</p>

<pre class="syntaxbox">{
  statement_1;
  statement_2;
  .
  .
  .
  statement_n;
}
</pre>

<h3 id="Example" name="Example"><strong>用例</strong></h3>

<p>ブロック文は一般に制御フロー文（例えば <code>if</code>、<code>for</code>、<code>while</code> など）で用いられます。</p>

<pre class="brush: js">while (x &lt; 10) {
  x++;
}
</pre>

<p>ここでは <code>{ x++; }</code> がブロック文となります。</p>

<p><strong>重要 :</strong> ECMAScript 2015 より前の JavaScript にはブロックスコープが<strong>ありません</strong>。ブロック内で導入された変数のスコープは、そのブロックがある関数やスクリプトになります。またそれらの変数を設定した影響は、そのブロックを越えて持続します。つまり、ブロック文はスコープを定義しないということです。「独立した」ブロックも正しい構文ですが、C や Java のブロックで提供されるものとは異なる結果をもたらします。例えば :</p>

<pre class="brush: js">var x = 1;
{
  var x = 2;
}
console.log(x); // 2 を出力
</pre>

<p>ブロック内の <code>var x</code> 文はブロックの前の <code>var x</code> 文と同じスコープ内にあるため、この例では 2 が出力されます。C や Java では、同様のコードで 1 が出力されます。</p>

<p>ECMAScript 2015 からは、<code>let</code> 文や <code>const</code> による変数宣言がブロックスコープとなります。詳しくは {{jsxref("Statements/let", "let")}} 文や {{jsxref("Statements/const", "const")}} の参考ページをご覧ください。</p>

<h2 id="Conditional_statements" name="Conditional_statements">条件文</h2>

<p>条件文は、指定した条件が true ならば実行される命令の集まりです。JavaScript は <code>if...else</code> と <code>switch</code> の 2 つの条件文をサポートしています。</p>

<h3 id="if...else_statement" name="if...else_statement"><code>if...else</code> 文</h3>

<p>論理条件が true の場合に、文を実行するには <code>if</code> 文を使用してください。オプションの <code>else</code> 節を使用すると、条件が false の場合にも文を実行することができます。<code>if</code> 文は次のように使用します :</p>

<pre class="syntaxbox">if (condition) {
  statement_1;
} else {
  statement_2;
}</pre>

<p><code>condition</code> には、true か false に評価される式が入ります。どういったものが <code>true</code> や <code>false</code> として評価されるかについては <a href="/ja/docs/Web/JavaScript/Reference/Global_Objects/Boolean#Description">Boolean オブジェクト</a>をご覧ください。<code>condition</code> が true に評価された場合は <code>statement_1</code> が実行され、そうでない場合は <code>statement_2</code> が実行されます。<code>statement_1</code> と <code>statement_2</code> はどのような文でもかまいません。<code>if</code> をさらに入れ子にすることもできます。</p>

<p>次のように <code>else if</code> を使用して文を重ねることで、複数の条件を順々にテストすることができます :</p>

<pre class="syntaxbox">if (condition_1) {
  statement_1;
} else if (condition_2) {
  statement_2;
} else if (condition_n) {
  statement_n;
} else {
  statement_last;
}
</pre>

<p>複数の文を実行するには、ブロック文 (<code>{ ... }</code>) を使用してその文をグループ化してください。一般に、常にブロック文を使用するのが優れた方法です。特に <code>if</code> 文を入れ子にしたコードで有効です :</p>

<pre class="syntaxbox">if (condition) {
  statement_1_runs_if_condition_is_true;
  statement_2_runs_if_condition_is_true;
} else {
  statement_3_runs_if_condition_is_false;
  statement_4_runs_if_condition_is_false;
}
</pre>

<p>条件式内で単純な代入を行わないでください。コードを一見した際に、代入を等値条件と見間違えるおそれがあるためです。例えば、次のコードにあげるような使用法はやめてください :</p>

<pre class="example-bad brush: js">if (x = y) {
  /* ここに文が来る */
}
</pre>

<p>条件式で代入を行う必要がある場合、一般的な方法は代入式をさらに丸括弧でくくることです。例えば :</p>

<pre class="brush: js">if ((x = y)) {
  /* ここに文が来る */
}
</pre>

<h4 id="Falsy_values" name="Falsy_values">false と評価される値</h4>

<p>以下の値は false に評価されます :</p>

<ul>
 <li><code>false</code></li>
 <li><code>undefined</code></li>
 <li><code>null</code></li>
 <li><code>0</code></li>
 <li><code>NaN</code></li>
 <li>空の文字列 (<code>""</code>)</li>
</ul>

<p>上記以外の、オブジェクトを含むすべての値は、条件文に渡されると true に評価されます。</p>

<p>プリミティブな真偽値の <code>true</code> と <code>false</code> を、{{jsxref("Boolean")}} オブジェクトの true や false という値と混同しないでください。例えば :</p>

<pre class="brush: js">var b = new Boolean(false);
if (b) // この条件は true に評価される
if (b == true) // この条件は false に評価される
</pre>

<h4 id="Example_2" name="Example_2">用例</h4>

<p>次の例で、関数 <code>checkData</code> は <code>Text</code> オブジェクトに含まれている文字数が 3 である場合に true を返し、そうでない場合はアラートを表示して false を返します。</p>

<pre class="brush: js">function checkData() {
  if (document.form1.threeChar.value.length == 3) {
    return true;
  } else {
    alert('Enter exactly three characters. ' +
    document.form1.threeChar.value + ' is not valid.');
    return false;
  }
}
</pre>

<h3 id="switch_statement" name="switch_statement"><code>switch</code> 文</h3>

<p><code>switch</code> 文を使うと、プログラムは式を評価し、その式の値を case ラベルと照合します。マッチすると、プログラムはそのラベルに関連付けられた文を実行します。<code>switch</code> 文は次のようになります :</p>

<pre class="syntaxbox">switch (expression) {
  case label_1:
    statements_1
    [break;]
  case label_2:
    statements_2
    [break;]
    ...
  default:
    statements_def
    [break;]
}
</pre>

<p>プログラムはまず、式の値にマッチするラベルを持つ <code>case</code> 節を探します。そして制御をその節に移し、関連付けられた文を実行します。マッチするラベルがない場合、プログラムはオプションの <code>default</code> 節を探し、存在する場合は <code>default</code> 節に制御を移し、関連付けられた文を実行します。<code>default</code> 節がない場合、プログラムは <code>switch</code> 文の最後まで、後に続く文を実行し続けます。慣例により、<code>default</code> 節は最後の節に置きますが、そうしなければいけないわけではありません。</p>

<p>各 <code>case</code> に関連付けられたオプションの <code>break</code> 文は、マッチした文を実行したら <code>switch</code> 文から抜けて、確実に switch 文以後の文を実行するためのものです。<code>break</code> を省略した場合、プログラムはその <code>switch</code> 文内の次の文から実行を続けます。</p>

<h4 id="Example_3" name="Example_3">用例</h4>

<p>次の例では、<code>fruittype</code> が "Bananas" に評価された場合は case "Bananas" にマッチして、それに関連付けられた文を実行します。<code>break</code> 文に出くわすとプログラムは <code>switch</code> から抜けて、<code>switch</code> の後に続く文を実行します。<code>break</code> を省略すると、case "Cherries" の文も実行されます。</p>

<pre class="brush: js">switch (fruittype) {
  case 'Oranges':
    console.log('Oranges are $0.59 a pound.');
    break;
  case 'Apples':
    console.log('Apples are $0.32 a pound.');
    break;
  case 'Bananas':
    console.log('Bananas are $0.48 a pound.');
    break;
  case 'Cherries':
    console.log('Cherries are $3.00 a pound.');
    break;
  case 'Mangoes':
    console.log('Mangoes are $0.56 a pound.');
    break;
  case 'Papayas':
    console.log('Mangoes and papayas are $2.79 a pound.');
    break;
  default:
   console.log('Sorry, we are out of ' + fruittype + '.');
}
console.log("Is there anything else you'd like?");</pre>

<h2 id="Exception_handling_statements" name="Exception_handling_statements">例外処理文</h2>

<p><code>throw</code> 文を使用して例外を発生させる（例外をスローする）ことができます。また、<code>try...catch</code> 文を使用して例外を処理することができます。</p>

<ul>
 <li><a href="#throw_statement"><code>throw</code> 文</a></li>
 <li><a href="#try_catch_statement"><code>try...catch</code> 文</a></li>
</ul>

<h3 id="Exception_types" name="Exception_types">例外の種類</h3>

<p>JavaScript では、ほぼどのようなオブジェクトでも例外としてスローすることができます。とはいえ、必ずしもスローされるオブジェクトすべてが同等に作られているわけではありません。数値や文字列をエラーとしてスローする方法がよく用いられますが、こうした用途のために特別に作られた例外データ型を使用した方がより効率的な場合もあります :</p>

<ul>
 <li><a href="/ja/docs/Web/JavaScript/Reference/Global_Objects#Fundamental_objects">ECMAScript 例外</a></li>
 <li>{{domxref("DOMException")}} と {{domxref("DOMError")}}</li>
</ul>

<h3 id="throw_statement" name="throw_statement"><code>throw</code> 文</h3>

<p><code>throw</code> 文は、例外をスローするために使用します。例外をスローするには、スローしたい値を含む式を指定してください :</p>

<pre class="syntaxbox">throw expression;
</pre>

<p>特定の型の式だけではなく、あらゆる式をスローすることができます。下記のコードでは、さまざまな型の例外をスローしています :</p>

<pre class="brush: js">throw "Error2";   // String type
throw 42;         // Number type
throw true;       // Boolean type
throw {toString: function() { return "I'm an object!"; } };
</pre>

<div class="note"><strong>注記 :</strong> 例外をスローする際にオブジェクトを指定することができます。そして、<code>catch</code> ブロックでそのオブジェクトのプロパティを参照できます。</div>

<pre class="brush: js">// UserException というオブジェクト型を作成
function UserException(message) {
  this.message = message;
  this.name = 'UserException';
}

// 文字列として使用されるとき（例 : エラーコンソール上）に
// 例外を整形する
UserException.prototype.toString = function() {
  return this.name + ': "' + this.message + '"';
}

// UserException のインスタンスを作成し、それをスローする
throw new UserException('Value too high');</pre>

<h3 id="try...catch_statement" name="try...catch_statement"><code>try...catch</code> 文</h3>

<p><code>try...catch</code> 文はテストしたい文のブロックを指定し、さらにスローされる例外に対する 1 つ以上の対処方法を指定します。例外がスローされると、<code>try...catch</code> 文がそれを受け取ります。</p>

<p><code>try...catch</code> 文は 1 つの <code>try</code> ブロックと 0 個以上の <code>catch</code> ブロックで構成されます。<code>try</code> ブロックは 1 つ以上の文を含み、<code>catch</code> ブロックは <code>try</code> ブロックで例外がスローされた場合の処理を指定する文が含まれます。要するに、成功した場合に実行したい <code>try</code> ブロックと、失敗した場合に制御を移行させたい <code>catch</code> ブロックで構成されています。<code>try</code> ブロック内（もしくは <code>try</code> ブロック内から呼び出された関数内）のいずれかの文が例外をスローすると、制御はすぐに <code>catch</code> ブロックに移ります。<code>try</code> ブロックで例外がスローされなかった場合、<code>catch</code> ブロックはスキップされます。<code>finally</code> ブロックは <code>try</code> および <code>catch</code> ブロックを実行した後に実行しますが、<code>try...catch</code> 文の後に続く文より先に実行されます。</p>

<p>次の例では <code>try...catch</code> 文を使用しています。この例では渡された値に基づいて、配列から月の名前を取り出す関数を実行します。値に対応する月の数字 (1-12) が無い場合は <code>"InvalidMonthNo"</code> という値を持つ例外がスローされ、<code>catch</code> ブロックの中の文は <code>monthName</code> という変数に <code>unknown</code> という値をセットします。</p>

<pre class="brush: js">function getMonthName(mo) {
  mo = mo - 1; // 月の数字を配列のインデックスに合わせる (1=Jan, 12=Dec)
  var months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul',
                'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
  if (months[mo]) {
    return months[mo];
  } else {
    throw 'InvalidMonthNo'; //throw キーワードが使われている
  }
}

try { // 実行を試みる文
  monthName = getMonthName(myMonth); // この関数が例外をスローする場合がある
}
catch (e) {
  monthName = 'unknown';
  logMyErrors(e); // 例外オブジェクトをエラーハンドラに渡す
}
</pre>

<h4 id="The_catch_Block" name="The_catch_Block"><code>catch</code> ブロック</h4>

<p><code>catch</code> ブロックを用いることで、<code>try</code> ブロックで生じうるすべての例外を扱うことができます。</p>

<pre class="syntaxbox">catch (catchID) {
  statements
}
</pre>

<p><code>catch</code> ブロックには、<code>throw</code> 文で指定される値を保持しておく識別子（上記の構文における <code>catchID</code>）を指定します。スローされた例外についての情報を得るのに、この識別子を使います。JavaScript は <code>catch</code> ブロックに入るときにこの識別子を作成します。識別子は <code>catch</code> ブロックの区間だけ存続します。つまり、<code>catch</code> ブロックの実行が終わると、その識別子はもう使えなくなります。</p>

<p>例えば、次のコードは例外をスローします。例外が生じると、制御が <code>catch</code> ブロックに移ります。</p>

<pre class="brush: js">try {
  throw 'myException'; // 例外を生成
}
catch (e) {
  // ここには例外を扱う文が入る
  logMyErrors(e); // 例外オブジェクトをエラーハンドラに渡す
}
</pre>

<h4 id="The_finally_block" name="The_finally_block"><code>finally</code> ブロック</h4>

<p><code>finally</code> ブロックは、<code>try</code> および <code>catch</code> ブロックの実行が終わった後に、<code>try...catch</code> 文の後に続く文よりも先に実行される文で構成されます。<code>finally</code> は例外がスローされてもされなくても実行されます。例外がスローされた場合、<code>finally</code> ブロック内の文はたとえ例外処理をする <code>catch</code> ブロックがなくても実行されます。</p>

<p><code>finally</code> ブロックを使用することで、例外発生時に適切にスクリプトを停止させることができます。例えば、スクリプトで使用していたリソースを解放しなければならないとします。次の例ではファイルを開き、そのファイルを使用する文を実行します (サーバーサイド JavaScript ではファイルにアクセスできます)。ファイルを開いている間に例外がスローされると、スクリプトが停止する前に <code>finally</code> ブロックでそのファイルを閉じます。</p>

<pre class="brush: js">openMyFile();
try {
  writeMyFile(theData); // ここでエラーがスローされる可能性がある
} catch(e) {  
  handleError(e); // エラーを受け取り、それを処理する
} finally {
  closeMyFile(); // 常にリソースが閉じられる
}
</pre>

<p><code>finally</code> ブロックが値を返す場合、その値は <code>try</code> および <code>catch</code> ブロックの <code>return</code> 文にかかわらず、<code>try-catch-finally</code> 全体が生成する戻り値になります :</p>

<pre class="brush: js">function f() {
  try {
    console.log(0);
    throw 'bogus';
  } catch(e) {
    console.log(1);
    return true; // この戻り値は、finally ブロックが
                 // 完了するまで保留となる
    console.log(2); // この先は実行されない
  } finally {
    console.log(3);
    return false; // 直前の "return" が上書きされる
    console.log(4); // この先は実行されない
  }
  // ここで "return false" が実行される
  console.log(5); // この先は実行されない
}
f(); // 0, 1, 3 がログに表示され、false が返される
</pre>

<p><code>finally</code> ブロックによる返値の上書きは、<code>catch</code> ブロック内で throw または再 throw された例外にも適用されます。</p>

<pre class="brush: js">function f() {
  try {
    throw 'bogus';
  } catch(e) {
    console.log('caught inner "bogus"');
    throw e; // この throw 文は finally ブロックが
             // 完了するまで保留になる
  } finally {
    return false; // 直前の "throw" が上書きされる
  }
  // ここで "return false" が実行される
}

try {
  f();
} catch(e) {
  // catch の中の throw が finally の中の
  // return で上書きされるので、
  // ここには決して到達しない
  console.log('caught outer "bogus"');
}

// OUTPUT
// caught inner "bogus"</pre>

<h4 id="Nesting_try...catch_Statements" name="Nesting_try...catch_Statements">try...catch 文のネスト</h4>

<p>1 つ以上の <code>try...catch</code> 文を入れ子にすることができます。内側の <code>try...catch</code> 文に <code>catch</code> ブロックがない場合、囲んでいる <code>try...catch</code> 文の <code>catch</code> ブロックがエラーの照合先としてチェックされます。詳しくは、<code><a href="/ja/docs/Web/JavaScript/Reference/Statements/try...catch">try...catch</a></code> の中の <a href="/ja/docs/Web/JavaScript/Reference/Statements/try...catch#Nested_try-blocks">nested try-blocks</a> を参照してください。</p>

<h3 id="Utilizing_Error_objects" name="Utilizing_Error_objects"><code>Error</code> オブジェクトの活用</h3>

<p>エラーの種類に応じてより手の込んだメッセージが得られるように、'name' および 'message' プロパティを使うことができます。'name' はエラーの一般的なクラス（例えば 'DOMException' や 'Error'）を表し、一方 'message' は通常、Error オブジェクトを文字列に変換したものより簡潔なメッセージを表します。</p>

<p>独自の例外をスローして、これらのプロパティを有効に活用したい場合（catch ブロックで独自の例外とシステムの例外とを区別したくない場合など）、Error コンストラクタが使えます。例えば :</p>

<pre class="brush: js">function doSomethingErrorProne() {
  if (ourCodeMakesAMistake()) {
    throw (new Error('The message'));
  } else {
    doSomethingToGetAJavascriptError();
  }
}
....
try {
  doSomethingErrorProne();
} catch (e) {
  console.log(e.name); // 'Error' とログに表示される
  console.log(e.message); // 'The message'、または
                          // JavaScript のエラーメッセージがログに表示される
}</pre>

<h2 id="Promises" name="Promises">Promise</h2>

<p>ECMAScript 2015 から、JavaScript には遅延・非同期処理の制御フローを可能にする {{jsxref("Promise")}} オブジェクトのサポートが追加されました。</p>

<p><code>Promise</code> は以下にある状態の一つを取ります :</p>

<ul>
 <li><em>pending</em> （処理待ち / 未決）: 初期状態で、承認も拒否もされていない状態。</li>
 <li><em>fulfilled</em> （承認 / 可決）: 処理が成功した状態。</li>
 <li><em>rejected</em> （拒否 / 否決）: 処理が失敗した状態。</li>
 <li><em>settled</em> （解決）: Promise オブジェクトが処理待ちでなく、承認か拒否された状態。</li>
</ul>

<p><img alt="" src="https://mdn.mozillademos.org/files/8633/promises.png" style="height: 297px; width: 801px;"></p>

<h3 id="Loading_an_image_with_XHR" name="Loading_an_image_with_XHR">XHR を使った画像のロード</h3>

<p>画像のロードに <code>Promise</code> と <code><a href="/ja/docs/Web/API/XMLHttpRequest">XMLHttpRequest</a></code> を使った分かりやすい例が MDN GitHub <a href="https://github.com/mdn/js-examples/tree/master/promises-test">js-examplest</a> リポジトリで利用できます。<a href="https://mdn.github.io/js-examples/promises-test/">実際の動き</a>も確認できます。それぞれのステップにはコメントが付けられ、Promise と XHR の基本概念をじっくりと観察することができます。下記にあるのはコメントのないバージョンですが、<code>Promise</code> の制御フローが手に取るようにわかると思います。:</p>

<pre class="brush: js">function imgLoad(url) {
  return new Promise(function(resolve, reject) {
    var request = new XMLHttpRequest();
    request.open('GET', url);
    request.responseType = 'blob';
    request.onload = function() {
      if (request.status === 200) {
        resolve(request.response);
      } else {
        reject(Error('Image didn\'t load successfully; error code:' 
                     + request.statusText));
      }
    };
    request.onerror = function() {
      reject(Error('There was a network error.'));
    };
    request.send();
  });
}</pre>

<p>詳細については、{{jsxref("Promise")}} オブジェクトのリファレンスページをご覧ください。</p>

<div>{{PreviousNext("Web/JavaScript/Guide/Grammar_and_types", "Web/JavaScript/Guide/Loops_and_iteration")}}</div>